data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ParseTreeEvaluatorListener.java

```diff
@@ -0,0 +1,387 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.expression;
+
+import com.amazon.dataprepper.model.event.Event;
+import org.antlr.v4.runtime.ParserRuleContext;
+import org.antlr.v4.runtime.Token;
+import org.antlr.v4.runtime.tree.ErrorNode;
+import org.antlr.v4.runtime.tree.TerminalNode;
+import org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionListener;
+import org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.Stack;
+
+import static org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser.AND;
+import static org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser.EQUAL;
+import static org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser.GT;
+import static org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser.GTE;
+import static org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser.LT;
+import static org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser.LTE;
+import static org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser.NOT;
+import static org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser.NOT_EQUAL;
+import static org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser.OR;
+
+/**
+ * @since 1.3
+ * This listener implements {@link DataPrepperExpressionListener} to provide callbacks to handle evaluation of
+ * {@link org.antlr.v4.runtime.tree.ParseTree} representation of an expression while {@link org.antlr.v4.runtime.tree.ParseTreeWalker}
+ * traverses through the {@link org.antlr.v4.runtime.tree.ParseTree}.
+ *
+ * Use case:
+ * ParseTreeWalker walker = new ParseTreeWalker();
+ * ParseTreeEvaluatorListener listener = new ParseTreeEvaluatorListener(...);
+ * listener.initialize(...);
+ * walker.walk(listener, ...);
+ * final Object result = listener.getResult();
+ */
+@Named
+class ParseTreeEvaluatorListener implements DataPrepperExpressionListener {
+    private static final Set<Integer> CONDITIONAL_OPERATOR_TYPES = Set.of(AND, OR);
```

Comment:
Having these sets is non-orthogonal. When adding new operators, you must come and modify this list.

I'm unsure of the best solution, but this one could work:

1. Create an enum to represent an "operator set".
2. Add `getOperatorSet` to the `Operator` interface.
3. Each `Operator` will implement this.
4. The `ParseTreeEvaluatorListener` can then double-check the advertised operator set against the expected.

Something like this:

```
enum OperatorSet {
  CONDITIONAL,
  EQUALITY,
  RELATIONAL
}

...

interface Operator {
  OperatorSet getOperatorSet();
}

class AndOperator {
  OperatorSet getOperatorSet() {
    return OperatorSet.CONDITIONAL;
  }
}
```